Stask 5

JugglypuffM opened the pull request.

Commit:
Concurrent linked list implemented and integrated into log source

Commit:
Merge branch 'master' into Stask-4

# Conflicts:
#	src/main/java/log/LogWindowSource.java

Commit:
underlying structure change

Commit:
multiple data extraction complexity reduced by replacing list with iterator

Commit:
log listeners unregistration on exit

Commit:
windows and menu localization

vpyzhyanov commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -21,6 +28,7 @@
 public class MainApplicationFrame extends JFrame implements Memorizable {
     private final JDesktopPane desktopPane = new JDesktopPane();
     private final StateManager stateManager = new StateManager();
+    public ResourceBundle bundle = ResourceBundle.getBundle("localization", new Locale("ru"));
```
то делает это поле в классе MainApplicationFrame? Ему тут явно не место. Это тёмная сторона ООП так-то, основной класс теперь ещё и за локализацию отвечает. Это должен делать отдельный класс

vpyzhyanov commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -70,14 +78,31 @@ protected void addWindow(JInternalFrame frame) {
         frame.setVisible(true);
     }
 
+    public void changeLocale(Locale locale) {
```
где JavaDoc?

vpyzhyanov commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -70,14 +78,31 @@ protected void addWindow(JInternalFrame frame) {
+                    if (icon.getInternalFrame() instanceof Localizable localizable)
+                        localizable.localeChange(bundle);
+        } catch (MissingResourceException e) {
+            Logger.error("Locale change failed due to missing resources with message:\n" + e.getMessage());
```
лучше стек тоже выводить, он будет очень полезен при разборе ошибок

vpyzhyanov commented on src/main/java/gui/menu/AppMenu.java:
```diff
@@ -0,0 +1,36 @@
+    private final static String CLASSNAME = "appMenu";
+
+    public AppMenu(MainApplicationFrame mainframe) {
+        super("Приложение");
```
А где локализация?

vpyzhyanov commented on src/main/java/gui/menu/AppMenu.java:
```diff
@@ -0,0 +1,36 @@
+    public AppMenu(MainApplicationFrame mainframe) {
+        super("Приложение");
+        setMnemonic(KeyEvent.VK_T);
+        getAccessibleContext().setAccessibleDescription("Команды приложению");
```
И тут нет

vpyzhyanov commented on src/main/java/gui/menu/LocaleMenu.java:
```diff
@@ -0,0 +1,36 @@
+    private final static String CLASSNAME = "localeMenu";
+
+    public LocaleMenu(MainApplicationFrame mainframe) {
+        super("Язык");
```
Это же новый класс, почему нет локализации?!

vpyzhyanov commented on src/main/java/gui/menu/TestMenu.java:
```diff
@@ -0,0 +1,30 @@
+        setMnemonic(KeyEvent.VK_T);
+        getAccessibleContext().setAccessibleDescription("Тестовые команды");
+        JMenuItem addLogMessageItem = new JMenuItem("Сообщение в лог", KeyEvent.VK_S);
+        addLogMessageItem.addActionListener((event) -> Logger.debug("Новая строка"));
```
нигде нет локализации, печаль( Это точно задача на локализацию?

JugglypuffM replied:
да я както не подумал, что значения по умолчания надо менять)

vpyzhyanov replied:
не исправлено!

vpyzhyanov replied:
опять не исправлено! Второй раз не исправляется замечание, притом без какого либо обоснования. За такие вещи вам полагается предупреждение! С последующим ущербом для репутации.

JugglypuffM replied:
Я не понимаю, что происходит, но это кажется какой-то баг - я не могу найти эти строки ни в коде локально ни в гитхабе.
Даже если во вкладке "изменённые файлы" если посмотреть на этот комментарий - он отображается рядом с другим кодом, не тем что выделен в комментарии

vpyzhyanov replied:
О_о почему снова не исправлено? В чём проблема?

JugglypuffM replied:
вроде наконец исправил)

vpyzhyanov reviewed (changes requested):
есть замечания

Commit:
mainframe change and components localization fix
localization responsibility moved from mainframe to locale manager and hardcoded component titles fixed

vpyzhyanov commented on src/main/java/gui/menu/LocaleMenu.java:
```diff
@@ -14,16 +15,17 @@
-    public LocaleMenu(MainApplicationFrame mainframe) {
-        super("Язык");
+    public LocaleMenu(LocaleManager localeManager) {
+        super("");
```
Так себе исправление. Если вдруг метод по смене локали не сработает то класс вообще останется без названия. Во всех остальных местах тоже.

Правильно будет использовать ResourceBundle **везде**, а метод localeChange вызывать только при смене локали (у него даже название говорит о том, что он вызывается при смене локали)

vpyzhyanov replied:
Не исправлено!!! =(((((

vpyzhyanov commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -118,6 +104,14 @@ else if (component instanceof JInternalFrame.JDesktopIcon icon)
+    }
+
+    public LocaleManager getLocaleManager() {
+        return localeManager;
```
нарушение инкапсуляции. Лучше бы `LocaleManager` сделали сиглтоном

vpyzhyanov commented on src/main/java/gui/menu/LocaleMenu.java:
```diff
@@ -14,16 +15,17 @@
+        JMenuItem english = new JMenuItem("");
+        english.addActionListener((event) -> localeManager.changeLocale(new Locale("en")));
         add(english);
+        localeChange(LocaleManager.getBundle());
```
метод не должен вызываться в конструторе

vpyzhyanov commented on src/main/java/gui/menu/TestMenu.java:
```diff
@@ -14,12 +16,12 @@ public class TestMenu extends JMenu implements Localizable {
+        JMenuItem addLogMessageItem = new JMenuItem("", KeyEvent.VK_S);
         addLogMessageItem.addActionListener((event) -> Logger.debug("Новая строка"));
         add(addLogMessageItem);
+        localeChange(LocaleManager.getBundle());
```
разве вы не чувствуете потерю ООП, когда пишете в куче мест абсолютно одинаковый код?! А надо бы уже чувствовать

vpyzhyanov commented on src/main/java/localization/LocaleManager.java:
```diff
@@ -0,0 +1,51 @@
+
+/**
+ * Locale manager class
+ * Has static method to get access to resource bundle
```
Какой кошмар. Помянем ООП(((

vpyzhyanov commented on src/main/java/localization/LocaleManager.java:
```diff
@@ -0,0 +1,51 @@
+ * Has static method to get access to resource bundle
+ */
+public class LocaleManager {
+    private static ResourceBundle bundle = ResourceBundle.getBundle("localization", Locale.getDefault());
```
фу, какая гадость это ваш `static`. Применён совершенно неуместно. Бедное ООП и принципы SOLID, нам будет вас не хватать(((

JugglypuffM replied:
Не очень понимаю какой именно принцип я нарушил, объясните пожалуйста

JugglypuffM replied:
static я сделал чтобы не прокидывать во все локализуемые классы ссылку на него - и это действительно не по ооп, но вы сами говорили, что оно может быть неудобным, и сняли запрет со статиков.

Вообще не отрицаю, что мог применить его неуместно, но не знаю как тут можно сделать лучше)

vpyzhyanov replied:
> Не очень понимаю какой именно принцип я нарушил, объясните пожалуйста

Это принцип открытой закрытости. Кстати про это было в лекциях у Александра Владимировича.

> static я сделал чтобы не прокидывать во все локализуемые классы ссылку на него - и это действительно не по ооп, но вы сами говорили, что оно может быть неудобным, и сняли запрет со статиков.

Я уже предложил удобный вариант - сделать синглтон) Вариант очень удобный, и в этом случае очень уместный

JugglypuffM replied:
я просто вообще не сразу(если честно только сейчас) понял, что именно вы предлагали)

vpyzhyanov commented on src/main/java/localization/LocaleManager.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static ResourceBundle getBundle() {
+        return bundle;
```
лучше бы тут был 
```java
public static String getString(String messageCode) {
        return getInstance().getBundle().getString(messageCode);
}
```

vpyzhyanov reviewed (changes requested):
не исправлено(

Commit:
static fix and second default localization fix

vpyzhyanov commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -85,10 +85,9 @@ protected void addWindow(JInternalFrame frame) {
-        String[] options = {bundle.getString("exitDialog.yes"), bundle.getString("exitDialog.no")};
-        int option = JOptionPane.showOptionDialog(this, bundle.getString("exitDialog.question"),
-                bundle.getString("exitDialog.title"), JOptionPane.YES_NO_CANCEL_OPTION, JOptionPane.PLAIN_MESSAGE,
+        String[] options = {ResourceBundle.getBundle("localization", localeManager.getCurrentLocale()).getString("exitDialog.yes"), ResourceBundle.getBundle("localization", localeManager.getCurrentLocale()).getString("exitDialog.no")};
```
кучу раз дублируется один и тот же вызов `ResourceBundle.getBundle("localization", localeManager.getCurrentLocale()).getString(...)`. В одном из ранее сделанных замечаний я предлагал как это сделать лучше.

vpyzhyanov commented on src/main/java/gui/menu/AppMenu.java:
```diff
@@ -19,14 +20,14 @@ public class AppMenu extends JMenu implements Localizable {
 
     public AppMenu(MainApplicationFrame mainframe) {
         super("");
+        setText(ResourceBundle.getBundle("localization", Locale.getDefault()).getString(String.format("%s.label", CLASSNAME)));
```
В этом случае лучше использовать конкатенацию, а не `String.format`, т.к. конкатенация работает гораздо быстрее, чем парсинг шаблона. А ещё она кешируется, а значение сформированное по шаблону - нет.

vpyzhyanov commented on src/main/java/gui/menu/TestMenu.java:
```diff
@@ -17,11 +18,11 @@ public class TestMenu extends JMenu implements Localizable {
 
     public TestMenu() {
         super("");
```
не исправлено(

vpyzhyanov commented on src/main/java/gui/windows/CoordinateWindow.java:
```diff
@@ -22,6 +23,7 @@ public class CoordinateWindow extends JInternalFrame implements PropertyChangeLi
 
     public CoordinateWindow(StateManager stateManager, GameModel model) {
         super("", true, true, true, true);
```
не исправлено(

vpyzhyanov commented on src/main/java/gui/windows/GameWindow.java:
```diff
@@ -11,13 +11,15 @@
     private final GameVisualizer m_visualizer;
 
     public GameWindow(StateManager stateManager, GameModel model) {
         super("", true, true, true, true);
```
тоже

vpyzhyanov commented on src/main/java/localization/LocaleManager.java:
```diff
@@ -45,7 +46,7 @@ else if (component instanceof JInternalFrame.JDesktopIcon icon)
-    public static ResourceBundle getBundle() {
-        return bundle;
+    public Locale getCurrentLocale() {
+        return currentLocale;
```
лучше бы менеджер локалей локализованные сообщения отдавал, а не текущую локаль. А то получается класс для управления локалью есть, но ей управляют все, но только не он(

vpyzhyanov reviewed (changes requested):
Старания почему-то не ко всем замечаниям применены(

Commit:
LocaleManager made singleton and third localization fix

vpyzhyanov commented on src/main/java/gui/menu/MenuBar.java:
```diff
@@ -12,6 +13,6 @@ public MenuBar(MainApplicationFrame mainframe) {
         add(new TestMenu());
         add(new AppMenu(mainframe));
-        add(new LocaleMenu(mainframe.getLocaleManager()));
+        add(new LocaleMenu((JDesktopPane) mainframe.getContentPane(), this));
```
любой cast делать опасно, ведь это потенциальный ClassCastException. Лучше просто mainframe передать, и дальше уже разбираться с тем что вам от него нужно.

JugglypuffM replied:
Каст перенес в LocaleMenu, но просто mainframe передавать не вижу смысла - menuBar в любом случае нужно передавть через this, тк как в mainframe он еще не инициализирован, а кроме него нужен только contentPane и как бы зачен нам целый mainframe

Commit:
TestMenu log message localization and cast fix

vpyzhyanov commented:
Задача принята
